File: 0039-combination-sum/0039-combination-sum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        def backtrack(c, t, i, curr, ans, arr):
            if curr == t:
                ans.append(list(arr))
                return
            if i == len(c) or curr > t:
                return
            arr.append(c[i])
            backtrack(c, t, i, curr + c[i], ans, arr)
            arr.pop()
            backtrack(c, t, i + 1, curr, ans, arr)

        ans = []
        backtrack(candidates, target, 0, 0, ans, [])
        return ans

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        rec(candidates.length - 1, target, new ArrayList<>(), ans, candidates);
        return ans;
    }

    public void rec(int i, int t, List<Integer> c, List<List<Integer>> ans, int[] candidates) {
        if (t == 0) {
            ans.add(new ArrayList<>(c)); 
            return;
        }

        if (t < 0 || i < 0) {
            return;
        }

        rec(i - 1, t, c, ans, candidates);
        c.add(candidates[i]);
        rec(i, t - candidates[i], c, ans, candidates);
        c.remove(c.size() - 1); 
}
}